Extend conversion API with option to inject virtio drivers (#13748)
canonical/lxd: doc/howto/import_machines_to_instances.md

* When creating a container, you must provide a disk or partition that contains the root file system for the container.
  For example, this could be the `/` root disk of the machine or container where you are running the tool.
* When creating a virtual machine, you must provide a bootable disk, partition, or an image in raw, QCOW, QCOW2, VDI, VHDX, or VMDK format.
  This means that just providing a file system is not sufficient, and you cannot create a virtual machine from a container that you are running.
  It is also not possible to create a virtual machine from the physical machine that you are using to do the migration, because the migration tool would be using the disk that it is copying.
  Instead, you could provide a bootable image, or a bootable partition or disk that is currently not in use.

The tool can also inject the required drivers into the image:

> * Use `lxd-migrate --conversion=format,virtio` to ensure the image is converted into the raw format and virtio drivers are injected during the conversion.
>   The conversion option `virtio` requires `virt-v2v-in-place` to be installed on the host where LXD server runs.

Reviewer: ````suggestion
* To convert the image into raw format and inject the VIRTIO drivers during the conversion, use the following command:

       lxd-migrate --conversion=format,virtio

  ```{note}
  The conversion option `virtio` requires `virt-v2v-in-place` to be installed on the host where the LXD server runs.
  ```
````
To make it a bit easier to scan and find the important information.

Author: Thanks.

> * For converting Windows images, additional drivers needs to be installed on the host:

Reviewer: ```suggestion
* For converting Windows images, you must install additional drivers on the host:
```

Author: Thanks. Done :)

>    * Install the `virtio-win` package or download the [`virtio-win.iso`](https://fedorapeople.org/groups/virt/virtio-win/direct-downloads/stable-virtio/virtio-win.iso) and place it in `/usr/share/virtio-win` directory.
>    * Download `pnp_wait.exe` and `rhsrvany.exe` and place them in `/usr/share/virt-tools/` directory.

Reviewer: ```suggestion
   * Install the `virtio-win` package or download the [`virtio-win.iso`](https://fedorapeople.org/groups/virt/virtio-win/direct-downloads/stable-virtio/virtio-win.iso) and place it in the `/usr/share/virtio-win` directory.
   * Download `pnp_wait.exe` and `rhsrvany.exe` and place them in the `/usr/share/virt-tools/` directory.
```

* When creating a container, you must provide a disk or partition that contains the root file system for the container.
  For example, this could be the `/` root disk of the machine or container where you are running the tool.
* When creating a virtual machine, you must provide a bootable disk, partition, or an image in raw, QCOW, QCOW2, VDI, VHDX, or VMDK format.
  This means that just providing a file system is not sufficient, and you cannot create a virtual machine from a container that you are running.
  It is also not possible to create a virtual machine from the physical machine that you are using to do the migration, because the migration tool would be using the disk that it is copying.
  Instead, you could provide a bootable image, or a bootable partition or disk that is currently not in use.

The tool can also inject the required drivers into the image:

* To convert the image into raw format and inject the VIRTIO drivers during the conversion, use the following command:

>     lxd-migrate --conversion=format,virtio

Reviewer: This needs more indentation (should be 6 spaces I think).

Author: Fixed. Thanks :)

  ```{note}
  The conversion option `virtio` requires `virt-v2v-in-place` to be installed on the host where the LXD server runs.
  ```

* For converting Windows images, you must additional drivers on the host:

>    * Install the `virtio-win` package or download the [`virtio-win.iso`](https://fedorapeople.org/groups/virt/virtio-win/direct-downloads/stable-virtio/virtio-win.iso) and place it in the `/usr/share/virtio-win` directory.

Reviewer: ```suggestion
   * Install the `virtio-win` package or download the [`virtio-win.iso`](https://fedorapeople.org/groups/virt/virtio-win/direct-downloads/stable-virtio/virtio-win.iso) file and place it in the `/usr/share/virtio-win` directory.
```

  It is also not possible to create a virtual machine from the physical machine that you are using to do the migration, because the migration tool would be using the disk that it is copying.
  Instead, you could provide a bootable image, or a bootable partition or disk that is currently not in use.

> The tool can also inject the required drivers into the image:

Reviewer 2: required *virtio* drivers

* To convert the image into raw format and inject the VIRTIO drivers during the conversion, use the following command:

      lxd-migrate --conversion=format,virtio

  ```{note}
  The conversion option `virtio` requires `virt-v2v-in-place` to be installed on the host where the LXD server runs.
  ```

* For converting Windows images from a foreign hypervisor (not from QEMU/KVM with Q35/`virtio-scsi`), you must install additional drivers on the host:
   * Install the `virtio-win` package or download the [`virtio-win.iso`](https://fedorapeople.org/groups/virt/virtio-win/direct-downloads/stable-virtio/virtio-win.iso) file and place it in the `/usr/share/virtio-win` directory.
   * Download [`rhsrvany.exe` and `pnp_wait.exe`](https://github.com/rwmjones/rhsrvany), and place them in the `/usr/share/virt-tools/` directory.

   ````{tip}
   If you want to convert a Windows VM from a foreign hypervisor manually,
   you must install both the required Windows drivers (as described above) and `virt-v2v` (>= 2.3.4).

   <details>

>    <summary>Expand to see how to convert your Windows VM using `virt-v2v`</summary>

Author: Thanks, probably this is also the reason why it detects `virt` as missing in dictionary.

>    <summary>Expand to see how to convert your Windows VM using `virt-v2v`</summary>

Reviewer: Yes - code style is automatically excluded from the spelling check, but since the markup didn't work here, it raised an error.

  It is also not possible to create a virtual machine from the physical machine that you are using to do the migration, because the migration tool would be using the disk that it is copying.
  Instead, you could provide a bootable image, or a bootable partition or disk that is currently not in use.

The tool can also inject the required drivers into the image:

* To convert the image into raw format and inject the VIRTIO drivers during the conversion, use the following command:

      lxd-migrate --conversion=format,virtio

  ```{note}
  The conversion option `virtio` requires `virt-v2v-in-place` to be installed on the host where the LXD server runs.
  ```

* For converting Windows images, you must install additional drivers on the host:
   * Install the `virtio-win` package or download the [`virtio-win.iso`](https://fedorapeople.org/groups/virt/virtio-win/direct-downloads/stable-virtio/virtio-win.iso) file and place it in the `/usr/share/virtio-win` directory.
   * Download `pnp_wait.exe` and `rhsrvany.exe` and place them in the `/usr/share/virt-tools/` directory.

   ````{tip}

>    If you want to convert a Windows VM from a foreign hypervisor manually (not from QEMU/KVM with Q35/`virtio-scsi`),

Author: @ru-fu 

The driver installation section (`Install the virtio-win package or download...`) repeats a few lines lower in the tip (`If you want to convert a Windows VM from a foreign hypervisor maunaly...`).

Could I just say something like this in the tip, to reduce the unnecessary repetition?

Tip title: `Expand to see how to integrate the required drivers to your Windows VM manually`
Tip content:
```
If you want to convert a Windows VM from a foreign hypervisor manually (not from 
QEMU/KVM with Q35/`virtio-scsi`), in addition to the required Windows drivers, you
must also install `virt-v2v`.

Using `virt-v2v`, you can convert images from a foreign hypervisor to `raw` images for
LXD and include the required drivers:

Example 1: ... "Rest of the content remains the same"
```

Reviewer: Good catch!

Yes, it makes sense to remove that part from the tip. Just make it very clear that you still need to follow the instructions above - something like ``If you want to convert a Windows VM from a foreign hypervisor manually (not from QEMU/KVM with Q35/`virtio-scsi`), you must install both the required Windows drivers (as described above) and `virt-v2v`.``
